### PR TITLE
Add config to disable autocast in AOTAutograd backward

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155237
* __->__ #155236

See https://github.com/pytorch/pytorch/issues/153044 for context.
This PR adds a new `assume_standard_autocast=False` config flag. When
flipped to True we disable autocast in AOTAutograd backwards.

In the next PR up I will attempt to turn this flag to True by default.

Test Plan:
- new test

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: torch/csrc/utils/python_dispatch.cpp [hunk 1]
```diff
@@ -699,6 +700,15 @@ void initDispatchBindings(PyObject* module) {
     return keys;
   });
   m.def("_dispatch_num_backends", []() { return c10::num_backends; });
+  m.def("_autocast_enabled_devices", []() {
+    std::vector<std::string> result;
+    for (auto device_type : at::autocast::kAllAutocastDeviceTypes) {
+      if (at::autocast::is_autocast_enabled(device_type)) {
+        result.push_back(c10::DeviceTypeName(device_type, /*lowercase*/ true));
```
Comment (bdhirsh) [inline torch/csrc/utils/python_dispatch.cpp:707]: It looks like the source-of-truth mapping from string to DeviceType is here: https://github.com/pytorch/pytorch/blob/main/c10/core/Device.cpp#L17. Should we try to re-use it?

pro == single source of truth is good

con == it's work to expose that source of truth to python properly, and also we haven't seen many new autocast devices in the last few years now that we've consolidated on PrivateUse1

File: aten/src/ATen/autocast_mode.h [hunk 2]
```diff
@@ -10,6 +10,16 @@
...
 namespace at::autocast {
 
+constexpr std::array<at::DeviceType, 7> kAllAutocastDeviceTypes = {
+    at::kCPU,
+    at::kCUDA,
+    at::kXPU,
+    at::kIPU,
+    at::kXLA,
+    at::kHPU,
+    at::kPrivateUse1,
```
Comment (bdhirsh) [inline aten/src/ATen/autocast_mode.h:20]: similarly, I see a similar list of "autocast devices" in this function: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/autocast_mode.h#L174

what do you think of either:

(1) figuring out how to expose that same list of devices to python so we have single source of truth

(2) Adding a `Note [Autocast Devices]` comment that we mirror to both call-sites, just so that all of the places in the code that need to change are easier to find if we don't decide to have one source of truth for the device list

File: test/functorch/test_aotdispatch.py [hunk 3]
```diff
@@ -620,6 +620,48 @@ def forward(self, x):
...
+            for p in model.parameters():
+                p.grad.data.zero_()
+
+            model = torch.compile(model, fullgraph=True, backend="aot_eager")
+            with torch.amp.autocast(device_type=device, enabled=True):
+                y = model(x)
+            loss = (y * 16384).sum()
+            loss.backward()
+            result = [p.grad.clone() for p in model.parameters()]
+            self.assertEqual(result, expected)
```
Comment (bdhirsh) [inline test/functorch/test_aotdispatch.py:657]: nice - the gradients here are properly different (and the test fails) when `assume_standard_autocast=False`?

**Review by bdhirsh — APPROVED** [review]
sgtm!

File: aten/src/ATen/autocast_mode.h (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/autocast_mode.h:20]: Yeah I'll do some more cleanup in this PR

File: torch/csrc/utils/python_dispatch.cpp [hunk 4]
```diff
@@ -699,6 +700,15 @@ void initDispatchBindings(PyObject* module) {
     return keys;
   });
   m.def("_dispatch_num_backends", []() { return c10::num_backends; });
+  m.def("_autocast_enabled_devices", []() {
```
Comment (albanD) [inline torch/csrc/utils/python_dispatch.cpp:703]: Why do you need this from C++? torch/amp/* should have all the enable checks in python already.

File: torch/_functorch/config.py [hunk 5]
```diff
@@ -229,6 +229,23 @@ def remote_autograd_cache_default() -> Optional[bool]:
 # of tensors in question.
 fake_tensor_propagate_real_tensors = False
 
+# If we assume that the user is using autocast in a standard way;
```
Comment (albanD) [inline torch/_functorch/config.py:232]: I'm not sure how "standard" that is tbh
A lot of people also enable autocast throughout their training loop (which leads to other issues and that's why we know haha).
Irrespective of the default, we should have a cristal clear error message here when our assumption is wrong and which setting the user should flip given their use of autocast.

Comment (zou3519) [reply torch/_functorch/config.py:232]: A crystal clear error message is BC-breaking.

The autocast docs say the following:
image

File: aten/src/ATen/autocast_mode.h (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/autocast_mode.h:20]: I'm not going to do the consolidation in this PR, because it has behavior changes. This is a pre-existing problem -- there was already a separate list of devices in one of the autocast files and that list was not comprehensive.

**Comment by zou3519** [comment]
hmm I"m going to open a new PR, this one is a bit difficult to save
